File: lib/js/src/Bigfloat.js
'use strict';

var Bigint = require("./Bigint.js");
var $$String = require("bs-platform/lib/js/string.js");
var Caml_array = require("bs-platform/lib/js/caml_array.js");
var Caml_format = require("bs-platform/lib/js/caml_format.js");

function mkBigfloat(v, d) {
  return {
          value: v,
          decimalPos: d
        };
}

var zero = {
  value: Bigint.zero,
  decimalPos: 0
};

var one = {
  value: Bigint.one,
  decimalPos: 0
};

var minusOne = {
  value: Bigint.minus_one,
  decimalPos: 0
};

function repeatString(n, s) {
  if (n === 0) {
    return "";
  } else {
    return s + repeatString(n - 1 | 0, s);
  }
}

var addCommas = (function (s) {
      var sp = s.split(".");
      var l = sp[0].replace(/(\d)(?=(\d\d\d)+(?!\d))/g, "$1,");
      if (sp.length > 1) {
        return l.concat(".", sp[1]);
      } else {
        return l;
      }
    });

var removeCommas = (function (s) {
      return s.replace(/,/g, "");
    });

function ofInt(i) {
  return {
          value: Bigint.of_int(i),
          decimalPos: 0
        };
}

function ofInt32(i) {
  return {
          value: Bigint.of_int32(i),
          decimalPos: 0
        };
}

function ofInt64(i) {
  return {
          value: Bigint.of_int64(i),
          decimalPos: 0
        };
}

function ofNativeint(i) {
  return {
          value: Bigint.of_nativeint(i),
          decimalPos: 0
        };
}

function ofString(string, decimalPos) {
  var s = removeCommas(string);
  var a = s.split(".");
  if (a.length === 1) {
    return {
            value: Bigint.mul(Bigint.of_string(s), Bigint.pow(Bigint.of_int(10), decimalPos)),
            decimalPos: decimalPos
          };
  } else if (a.length === 2) {
    var leftSide = Caml_array.caml_array_get(a, 0);
    var rightSide = Caml_array.caml_array_get(a, 1);
    var rightSideLength = rightSide.length;
    var rr = rightSideLength === decimalPos ? rightSide : (
        rightSideLength < decimalPos ? rightSide + repeatString(decimalPos - rightSideLength | 0, "0") : $$String.sub(rightSide, 0, decimalPos)
      );
    return {
            value: Bigint.of_string(leftSide + rr),
            decimalPos: decimalPos
          };
  } else {
    return ;
  }
}

function ofFloat(f, decimalPos) {
  var s = f.toString();
  var a = s.split(".");
  if (a.length === 1) {
    return {
            value: Bigint.of_float(f),
            decimalPos: decimalPos
          };
  } else if (a.length === 2) {
    var leftSide = Caml_array.caml_array_get(a, 0);
    var rightSide = Caml_array.caml_array_get(a, 1);
    var rightSideLength = rightSide.length;
    var rr = rightSideLength === decimalPos ? rightSide : (
        rightSideLength < decimalPos ? rightSide + repeatString(decimalPos - rightSideLength | 0, "0") : $$String.sub(rightSide, 0, decimalPos)
      );
    return {
            value: Bigint.of_string(leftSide + rr),
            decimalPos: decimalPos
          };
  } else {
    return {
            value: Bigint.zero,
            decimalPos: decimalPos
          };
  }
}

function toInt(t) {
  return Bigint.to_int(Bigint.div(t.value, Bigint.pow(Bigint.of_int(10), t.decimalPos)));
}

function toInt32(t) {
  return Bigint.to_int32(Bigint.div(t.value, Bigint.pow(Bigint.of_int(10), t.decimalPos)));
}

function toInt64(t) {
  return Bigint.to_int64(Bigint.div(t.value, Bigint.pow(Bigint.of_int(10), t.decimalPos)));
}

function toNativeint(t) {
  return Bigint.to_nativeint(Bigint.div(t.value, Bigint.pow(Bigint.of_int(10), t.decimalPos)));
}

function toString(t) {
  var s = Bigint.to_string(t.value);
  var sl = s.length;
  if (t.decimalPos === 0) {
    return s;
  } else if (sl <= t.decimalPos) {
    return "0." + (repeatString(t.decimalPos - sl | 0, "0") + s);
  } else {
    var r = s.slice(sl - t.decimalPos | 0, sl);
    var l = s.slice(0, sl - t.decimalPos | 0);
    return l + ("." + r);
  }
}

function toStringWithCommas(t) {
  return addCommas(toString(t));
}

function toFloat(t) {
  return Caml_format.caml_float_of_string(toString(t));
}

function abs(t) {
  return {
          value: Bigint.abs(t.value),
          decimalPos: t.decimalPos
        };
}

function neg(t) {
  return {
          value: Bigint.neg(t.value),
          decimalPos: t.decimalPos
        };
}

function add(x, y) {
  if (x.decimalPos === y.decimalPos) {
    return {
            value: Bigint.add(x.value, y.value),
            decimalPos: x.decimalPos
          };
  } else if (x.decimalPos > y.decimalPos) {
    var decimalPos = x.decimalPos - y.decimalPos | 0;
    return {
            value: Bigint.add(x.value, Bigint.mul(y.value, Bigint.pow(Bigint.of_int(10), decimalPos))),
            decimalPos: x.decimalPos
          };
  } else {
    var decimalPos$1 = y.decimalPos - x.decimalPos | 0;
    return {
            value: Bigint.add(y.value, Bigint.mul(x.value, Bigint.pow(Bigint.of_int(10), decimalPos$1))),
            decimalPos: y.decimalPos
          };
  }
}

function sub(x, y) {
  if (x.decimalPos === y.decimalPos) {
    return {
            value: Bigint.sub(x.value, y.value),
            decimalPos: x.decimalPos
          };
  } else if (x.decimalPos > y.decimalPos) {
    var decimalPos = x.decimalPos - y.decimalPos | 0;
    return {
            value: Bigint.sub(x.value, Bigint.mul(y.value, Bigint.pow(Bigint.of_int(10), decimalPos))),
            decimalPos: x.decimalPos
          };
  } else {
    var decimalPos$1 = y.decimalPos - x.decimalPos | 0;
    return {
            value: Bigint.sub(y.value, Bigint.mul(x.value, Bigint.pow(Bigint.of_int(10), decimalPos$1))),
            decimalPos: y.decimalPos
          };
  }
}

function mul(x, y) {
  return {
          value: Bigint.mul(x.value, y.value),
          decimalPos: x.decimalPos + y.decimalPos | 0
        };
}

function elongateBy(x, decimalPos) {
  return {
          value: Bigint.mul(x.value, Bigint.pow(Bigint.of_int(10), decimalPos)),
          decimalPos: x.decimalPos
        };
}

function adjustDecimalPos(x, y) {
  if (x.decimalPos === y.decimalPos) {
    return /* tuple */[
            x,
            y
          ];
  } else if (x.decimalPos > y.decimalPos) {
    var decimalPos = x.decimalPos - y.decimalPos | 0;
    return /* tuple */[
            x,
            {
              value: Bigint.mul(y.value, Bigint.pow(Bigint.of_int(10), decimalPos)),
              decimalPos: x.decimalPos
            }
          ];
  } else {
    var decimalPos$1 = y.decimalPos - x.decimalPos | 0;
    return /* tuple */[
            {
              value: Bigint.mul(x.value, Bigint.pow(Bigint.of_int(10), decimalPos$1)),
              decimalPos: y.decimalPos
            },
            y
          ];
  }
}

function div(x, y) {
  if (x.decimalPos === 0 && y.decimalPos === 0) {
    return {
            value: Bigint.div(x.value, y.value),
            decimalPos: x.decimalPos
          };
  } else {
    var match = adjustDecimalPos(x, y);
    var y$1 = match[1];
    var x$1 = match[0];
    var xFirstNonZero = Bigint.firstNonZero(x$1.value);
    var yFirstNonZero = Bigint.firstNonZero(y$1.value);
    if (xFirstNonZero !== undefined && yFirstNonZero !== undefined) {
      var yFirstNonZero$1 = yFirstNonZero;
      var xFirstNonZero$1 = xFirstNonZero;
      if (xFirstNonZero$1 >= x$1.decimalPos && yFirstNonZero$1 >= y$1.decimalPos) {
        return elongateBy({
                    value: Bigint.div(x$1.value, y$1.value),
                    decimalPos: x$1.decimalPos
                  }, x$1.decimalPos);
      } else {
        var reposition = xFirstNonZero$1 > yFirstNonZero$1 ? y$1.decimalPos - yFirstNonZero$1 | 0 : x$1.decimalPos - xFirstNonZero$1 | 0;
        var x$2 = elongateBy(x$1, reposition);
        var y$2 = elongateBy(y$1, reposition);
        return elongateBy({
                    value: Bigint.div(x$2.value, y$2.value),
                    decimalPos: x$2.decimalPos
                  }, x$2.decimalPos);
      }
    } else {
      return {
              value: Bigint.zero,
              decimalPos: x$1.decimalPos
            };
    }
  }
}

function compare(x, y) {
  var match = adjustDecimalPos(x, y);
  return Bigint.compare(match[0].value, match[1].value);
}

function equal(x, y) {
  return compare(x, y) === 0;
}

function leq(x, y) {
  return compare(x, y) < 1;
}

function geq(x, y) {
  return compare(x, y) > -1;
}

function lt(x, y) {
  return compare(x, y) < 0;
}

function gt(x, y) {
  return compare(x, y) > 0;
}

function leqZero(x) {
  return Bigint.compare(x.value, Bigint.zero) <= 0;
}

function geqZero(x) {
  return Bigint.compare(x.value, Bigint.zero) >= 0;
}

function ltZero(x) {
  return Bigint.compare(x.value, Bigint.zero) < 0;
}

function gtZero(x) {
  return Bigint.compare(x.value, Bigint.zero) > 0;
}

function min(x, y) {
  if (compare(x, y) < 0) {
    return y;
  } else {
    return x;
  }
}

function max(x, y) {
  if (compare(x, y) < 0) {
    return x;
  } else {
    return y;
  }
}

exports.mkBigfloat = mkBigfloat;
exports.zero = zero;
exports.one = one;
exports.minusOne = minusOne;
exports.ofInt = ofInt;
exports.ofInt32 = ofInt32;
exports.ofInt64 = ofInt64;
exports.ofNativeint = ofNativeint;
exports.ofFloat = ofFloat;
exports.ofString = ofString;
exports.toInt = toInt;
exports.toInt32 = toInt32;
exports.toInt64 = toInt64;
exports.toNativeint = toNativeint;
exports.toFloat = toFloat;
exports.toString = toString;
exports.toStringWithCommas = toStringWithCommas;
exports.adjustDecimalPos = adjustDecimalPos;
exports.abs = abs;
exports.neg = neg;
exports.add = add;
exports.sub = sub;
exports.mul = mul;
exports.div = div;
exports.compare = compare;
exports.equal = equal;
exports.leq = leq;
exports.geq = geq;
exports.lt = lt;
exports.gt = gt;
exports.leqZero = leqZero;
exports.geqZero = geqZero;
exports.ltZero = ltZero;
exports.gtZero = gtZero;
exports.min = min;
exports.max = max;
/* No side effect */

File: src/Bigfloat.ml
type t = {
  value: Bigint.t;
  decimals: int;
  }

let mkBigfloat v d = { value = v; decimals = d } 
       
let zero = { value = Bigint.zero; decimals = 0 }

let one = { value = Bigint.one; decimals = 0 }

let minusOne = { value = Bigint.minus_one; decimals = 0 }

(** Internal utility functions *)             
let rec repeatString n s = if n = 0 then "" else s ^ (repeatString (n - 1) s)

let addCommas: string -> string =
  [%bs.raw
    {|
    function (s) {
      var sp = s.split(".");
      var l = sp[0].replace(/(\d)(?=(\d\d\d)+(?!\d))/g, "$1,");
      if (sp.length > 1) {
        return l.concat(".", sp[1]);
      } else {
        return l;
      }
    }
     |}]

let removeCommas: string -> string =
  [%bs.raw
    {|
    function (s) {
      return s.replace(/,/g, "");
    }
   |}]

(** of conversions *)

let ofInt i =
  { value = (Bigint.of_int i); decimals = 0 }
  
let ofInt32 i =
  { value = (Bigint.of_int32 i); decimals = 0 }
  
let ofInt64 i =
  { value = (Bigint.of_int64 i); decimals = 0 }

let ofNativeint i =
  { value = (Bigint.of_nativeint i); decimals = 0 }

let ofString (string : string) decimals =
  let s = removeCommas string in
  let a = Js.String.split "." s in
  if Array.length a = 1
  then
    Belt.Result.Ok
      { value = (Bigint.mul (Bigint.of_string s) (Bigint.pow (Bigint.of_int 10) decimals));
        decimals }
   else
     if Array.length a = 2
     then
       let leftSide = a.(0) in
       let rightSide = a.(1) in
       let rightSideLength = String.length rightSide in
       let rr =
         (if rightSideLength = decimals
          then rightSide
          else
            if rightSideLength < decimals
            then
              rightSide ^
                (repeatString (decimals - rightSideLength) "0")
            else
              String.sub rightSide 0 decimals) in
       Belt.Result.Ok { value = (Bigint.of_string (leftSide ^ rr)); decimals }
     else Belt.Result.Error string
  
let ofFloat (f : float) decimals =
  (let s = Js.Float.toString f in
   let a = Js.String.split ((".")[@reason.raw_literal "."]) s in
   if (Array.length a) = 1
   then { value = (Bigint.of_float f); decimals }
   else
     if (Array.length a) = 2
     then
       (let leftSide = a.(0) in
        let rightSide = a.(1) in
        let rightSideLength = String.length rightSide in
        let rr =
          if rightSideLength = decimals
          then rightSide
          else
            if rightSideLength < decimals
            then
              rightSide ^
                (repeatString (decimals - rightSideLength)
                   (("0")[@reason.raw_literal "0"]))
            else String.sub rightSide 0 decimals in
        { value = (Bigint.of_string (leftSide ^ rr)); decimals })
     else { value = Bigint.zero; decimals } : t)


(** to conversions *)

let toInt t =
   Bigint.to_int (Bigint.div t.value (Bigint.pow (Bigint.of_int 10) t.decimals))
  
let toInt32 t =
   Bigint.to_int32 (Bigint.div t.value (Bigint.pow (Bigint.of_int 10) t.decimals))
  
let toInt64 t =
   Bigint.to_int64 (Bigint.div t.value (Bigint.pow (Bigint.of_int 10) t.decimals))

let toNativeint t =
   Bigint.to_nativeint (Bigint.div t.value (Bigint.pow (Bigint.of_int 10) t.decimals))
  
let toString (t : t) =
  let s = Bigint.to_string t.value in
  let sl = String.length s in
  if t.decimals = 0
  then s
  else
    if sl <= t.decimals
    then
      "0." ^ (repeatString (t.decimals - sl) "0") ^ s
    else
      (let r = Js.String.slice ~from:(sl - t.decimals) ~to_:sl s in
       let l = Js.String.slice ~from:0 ~to_:(sl - t.decimals) s in
       l ^ "." ^ r)

let toStringWithCommas (t : t) = (addCommas (toString t) : string)
  
let toFloat (t : t) = (float_of_string (toString t) : float)

(** Arithmetic *)

let abs t = {t with value = Bigint.abs t.value}

let neg t = {t with value = Bigint.neg t.value}
  
let add (x : t) (y : t) =
  if x.decimals = y.decimals
  then { value = (Bigint.add x.value y.value); decimals = (x.decimals) }
  else
    if x.decimals > y.decimals
    then
      (let decimals = x.decimals - y.decimals in
       {
         value =
           (let open Bigint in
              add x.value (mul y.value (pow (of_int 10) decimals)));
         decimals = (x.decimals)
       })
    else
      (let decimals = y.decimals - x.decimals in
       {
         value =
           (let open Bigint in
              add y.value (mul x.value (pow (of_int 10) decimals)));
         decimals = (y.decimals)
       })

let sub x y =
  if x.decimals = y.decimals
  then { value = (Bigint.sub x.value y.value); decimals = (x.decimals) }
  else
    if x.decimals > y.decimals
    then
      (let decimals = x.decimals - y.decimals in
       {
         value =
           (let open Bigint in
              sub x.value (mul y.value (pow (of_int 10) decimals)));
         decimals = (x.decimals)
       })
    else
      (let decimals = y.decimals - x.decimals in
       {
         value =
           (let open Bigint in
              sub y.value (mul x.value (pow (of_int 10) decimals)));
         decimals = (y.decimals)
       })

let mul x y =
  {
    value = (Bigint.mul x.value y.value);
    decimals = (x.decimals + y.decimals)
  }

let elongateBy x decimals =
  {
    x with
    value =
      (let open Bigint in Bigint.mul x.value (pow (of_int 10) decimals))
  }

(* let truncate x decimals =
 *   let diff = x.decimals - decimals in
 *   { value = (let open Bigint in div x.value (pow (of_int 10) diff)); decimals
 *   } *)

let adjustDecimals (x : t) (y : t) =
  if x.decimals = y.decimals
  then (x, y)
  else
    if x.decimals > y.decimals
    then
      (let decimals = x.decimals - y.decimals in
       (x,
         {
           value =
             (Bigint.mul y.value (Bigint.pow (Bigint.of_int 10) decimals));
           decimals = (x.decimals)
         }))
    else
      (let decimals = y.decimals - x.decimals in
       ({
          value = (Bigint.mul x.value (Bigint.pow (Bigint.of_int 10) decimals));
          decimals = (y.decimals)
        }, y))

let div x y =
  if (x.decimals = 0) && (y.decimals = 0)
  then { value = (Bigint.div x.value y.value); decimals = (x.decimals) }
  else
    (let (x,y) = adjustDecimals x y in
     let xFirstNonZero = Bigint.firstNonZero x.value in
     let yFirstNonZero = Bigint.firstNonZero y.value in
     match (xFirstNonZero, yFirstNonZero) with
     | (((Some (xFirstNonZero))[@explicit_arity ]),((Some
        (yFirstNonZero))[@explicit_arity ])) ->
         if (xFirstNonZero >= x.decimals) && (yFirstNonZero >= y.decimals)
         then
           elongateBy
             { value = (Bigint.div x.value y.value); decimals = (x.decimals)
             } x.decimals
         else
           (let reposition =
              if xFirstNonZero > yFirstNonZero
              then y.decimals - yFirstNonZero
              else x.decimals - xFirstNonZero in
            let x = elongateBy x reposition in
            let y = elongateBy y reposition in
            elongateBy
              {
                value = (let open Bigint in div x.value y.value);
                decimals = (x.decimals)
              } x.decimals)
     | _ -> { x with value = Bigint.zero })

(** comparison *)
  
let compare x y =
  let (x,y) = adjustDecimals x y in Bigint.compare x.value y.value

let equal x y = (compare x y) = 0

let leq x y = (compare x y) < 1

let geq x y = (compare x y) > -1
            
let lt x y = (compare x y) < 0

let gt x y = (compare x y) > 0
           
let leqZero x = (Bigint.compare x.value Bigint.zero) <= 0

let geqZero x = (Bigint.compare x.value Bigint.zero) >= 0
              
let ltZero x = (Bigint.compare x.value Bigint.zero) < 0

let gtZero x = (Bigint.compare x.value Bigint.zero) > 0

let min x y = (if (compare x y) < 0 then y else x : t)

let max x y = (if (compare x y) < 0 then x else y : t)
